Feature/rgcn

kieranricardo opened the pull request:
#164
Implementation of RGCN

#### Reviewer Checklist
- [x] Code correctly implements R-GCN algorithm
- [x] Unit tests
- [x] Demo notebook runs and clearly demonstrates/explains how to train R-GCN models
- [x] Documentation

PantelisElinas commented on stellargraph/layer/rgcn.py:
```diff
@@ -0,0 +1,496 @@
+        feat_shape = input_shapes[0]
+        input_dim = int(feat_shape[-1])
+
+        if self.num_bases != 0:
```
I guess negative values are good here!

Maybe change it to self.num_bases > 0 so that at least negative values are treated as zero. A check for a valid value and a unit test to go with it would be helpful.

kieranricardo replied:
Nice pickup! I'll change it to `self.num_bases > 0` and add a unit test around this

PantelisElinas commented on stellargraph/layer/rgcn.py:
```diff
@@ -0,0 +1,496 @@
+        for i in range(self.num_relationships):
+            h_graph = K.dot(As[i], features)
+            output += K.dot(h_graph, self.relational_kernels[i])
+
```
Either the above lines are incorrect or I have misunderstood the code or the paper. 

If I am not mistaken, according to equations 2 and 4 in the paper, there is a weight matrix, `W_r`, for each relationship type. In the above code, there is only one W, `self.self_kernel`. Isn't this wrong? In the paper's figure 2, they talk about parameter sharing but I don't think they mean sharing across relationships but rather across nodes. I may be wrong, but can you please check?

PantelisElinas commented on stellargraph/layer/rgcn.py:
```diff
@@ -0,0 +1,496 @@
+
+
+        where N is the number of nodes, F the number of input features,
+              E is the number of edges, O the number of output nodes,
```
Can you please use another symbol instead of O? When I first saw it, I thought it was a zero. Maybe use Z?

PantelisElinas commented on stellargraph/layer/rgcn.py:
```diff
@@ -0,0 +1,496 @@
+
+        return h_layer
+
+    def node_model(self):
```
We are trying to standardise on replacing `node_model()` with `build()`. The goal is to make `build()` return the correct model, node or link, based on the generator. Using `build()` will require the user to change fewer lines of code to re-use their scripts between tasks. Note that not all models have `build()` implemented yet, but we want to provide this method for all new model classes.

kieranricardo replied:
Should I use `build() `to call `node_model()` is the generator is node generator or removing the `node_model()` function entirely?

PantelisElinas replied:
I think it might be best to remove node_model all together by just making it a "private" method we can call from build().

kieranricardo replied:
done!

PantelisElinas commented on stellargraph/mapper/node_mappers.py:
```diff
@@ -1128,3 +1128,275 @@ def flow_from_dataframe(self, node_targets, shuffle=False):
+
+class RelationalSparseFullBatchNodeSequence(Sequence):
+    """
+    Keras-compatible data generator for for node inference models
```
Spelling?

PantelisElinas commented on stellargraph/mapper/node_mappers.py:
```diff
@@ -1128,3 +1128,275 @@ def flow_from_dataframe(self, node_targets, shuffle=False):
+        return self.inputs, self.targets
+
+
+class RelationalDenseFullBatchNodeSequence(Sequence):
```
This class looks nearly identical to the Sparse version above with 3-4 lines different to handle the different types of adjacency matrices.

Is there are a reason why we cannot combine them into a single class with a `sparse` parameter with default value `True`?

kieranricardo replied:
Great idea, I was copying the API for the `FullBatchNodeSequence` but this sounds much better. I'll implement this.

PantelisElinas commented on stellargraph/mapper/node_mappers.py:
```diff
@@ -1128,3 +1128,275 @@ def flow_from_dataframe(self, node_targets, shuffle=False):
+
+    def __init__(self, G, name=None, sparse=True, transform=None):
+
+        if not isinstance(G, StellarGraphBase):
```
In the docstring above, it says that G should be StellarDiGraph, but here the check if for the more general StellarGraph. Shall we update this test to be a bit more strict?

kieranricardo replied:
The paper only talks about using directed graphs (I think relational graphs are inherently directed?) but I don't see any reason why this wouldn't work for undirected graphs. I'll update the docstring

PantelisElinas commented on stellargraph/mapper/node_mappers.py:
```diff
@@ -1128,3 +1128,275 @@ def flow_from_dataframe(self, node_targets, shuffle=False):
+
+        # The list of indices of the target nodes in self.node_list
+        # use dictionary for faster index look-up time
+        node_index = dict(zip(self.node_list, range(len(self.node_list))))
```
You perform the same operation in the `__init__` method. Can you do it once and store it with the object in `self.node_index`? Or are you trying to save memory by using a temporary object?

kieranricardo replied:
I was trying to save memory, but I don't think time/memory overhead will be very much either. What do you think?

PantelisElinas replied:
Let's just do it in __init__. It will be easier to move it out of __init__ later, especially after we have re-worked the StellarGraph class and we have a better memory management solution.

kieranricardo replied:
done!

PantelisElinas commented on tests/layer/test_rgcn.py:
```diff
@@ -0,0 +1,322 @@
+    assert conf["use_bias"] == True
+    assert conf["kernel_initializer"]["class_name"] == "GlorotUniform"
+    assert conf["bias_initializer"]["class_name"] == "Zeros"
+    assert conf["kernel_regularizer"] == None
```
I think comparing for equality with None should use `is` not `==`.

kieranricardo replied:
You're absolutely right, fixed!

PantelisElinas commented on tests/layer/test_rgcn.py:
```diff
@@ -0,0 +1,322 @@
+    )
+
+    assert rgcn_layer.units == 16
+    assert rgcn_layer.use_bias == True
```
Shall we be using `is` here instead of `==`?

PantelisElinas commented on tests/mapper/test_relational_node_mappers.py:
```diff
@@ -0,0 +1,191 @@
+    node_features = pd.DataFrame.from_dict(
+        {n: f for n, f in zip(nodes, features)}, orient="index"
+    )
+    G = StellarDiGraph(gnx, node_features=node_features)
```
Please see my previous comments about the correct type of the graph input to RelationalFullBatchNodeGenerator. Does the input graph have to be directed or can it also be undirected? You only have a test for directed graphs.

kieranricardo replied:
It can be both, I've added in unit tests for both directed and undirected

PantelisElinas commented on tests/mapper/test_relational_node_mappers.py:
```diff
@@ -0,0 +1,191 @@
+
+    def test_generator_constructor_wrong_G_type(self):
+        with pytest.raises(TypeError):
+            generator = RelationalFullBatchNodeGenerator(nx.Graph(self.G))
```
Can we add tests for other graph types? Should nx.DiGraph or nx.MultidiGraph (the type of nx graph you start with in `create_graph_features` work?

kieranricardo replied:
This test is for making sure that passing any networkx graph to the generator will fail, only `StellarGraphBase` objects should be accepted.

adocherty commented on stellargraph/layer/rgcn.py:
```diff
@@ -0,0 +1,554 @@
+
+        if self.num_bases > 0:
+
+            self.bases = self.add_weight(
```
Add some more comments on what this code block is doing

kieranricardo replied:
Done!

adocherty commented on stellargraph/layer/rgcn.py:
```diff
@@ -0,0 +1,554 @@
+                    initializer=self.kernel_initializer,
+                    name="coeff",
+                    regularizer=self.kernel_regularizer,
+                    constraint=self.kernel_constraint,
```
I'm not sure that using the same regulariser for the coefficents as the kernel is what's required. I'd not include them in the coefficients.

kieranricardo replied:
good point: I'll create different regularisers, initialisers, and constraints for the kernels, coefficients, and bases

adocherty commented on stellargraph/layer/rgcn.py:
```diff
@@ -0,0 +1,554 @@
+        self.bias = bias
+        self.num_bases = num_bases
+        self.dropout = dropout
+        self.generator = generator
```
I'm trying to move away from storing the generator in the model. It seems too messy.

I have also introduced a 'multiplicity' property to the generators which is 1 for a node inference and 2 for link inference. To match the new-style full batch generators you could look at that.

See this PR: https://github.com/stellargraph/stellargraph/pull/543

kieranricardo replied:
yeah agreed, your new method is way nicer - done!

adocherty commented on stellargraph/layer/rgcn.py:
```diff
@@ -0,0 +1,554 @@
+        self.num_bases = num_bases
+        self.dropout = dropout
+        self.generator = generator
+        self.support = 1
```
This is not used in the code, remove?

kieranricardo replied:
done!

adocherty commented on stellargraph/layer/rgcn.py:
```diff
@@ -0,0 +1,554 @@
+
+        """
+
+        if self.generator is not None:
```
Look at https://github.com/stellargraph/stellargraph/pull/543 and change this to use the `multiplicity` flag to match the new-style full batch generators (don't implement link prediction yet though, but it would be a good idea to make a new issue to do this)

kieranricardo replied:
done!

adocherty commented on stellargraph/mapper/full_batch_generators.py:
```diff
@@ -238,3 +238,209 @@ def flow(self, node_ids, targets=None):
             )
+
+
+class RelationalFullBatchNodeSequence(Sequence):
```
Move this class to the `mapper/sequences.py` file.

kieranricardo replied:
done!

adocherty commented on stellargraph/mapper/full_batch_generators.py:
```diff
@@ -238,3 +238,209 @@ def flow(self, node_ids, targets=None):
+    """
+    A data generator for use with full-batch models on relational graphs e.g. RGCN.
+
+    The supplied graph G should be a StellarGraphBase object that is ready for
```
It should be a  `StellarGraph` or `StellarDiGraph` object. `StellarGraphBase` is the abstract base class for these objects and users can't create one.

kieranricardo replied:
good point, G can be either `StellarGraph` or `StellarDiGraph` so I'll change the docstring to say this

adocherty commented on stellargraph/mapper/full_batch_generators.py:
```diff
@@ -238,3 +238,209 @@ def flow(self, node_ids, targets=None):
+            A = sps.coo_matrix(
+                (data, (row_index, col_index)),
+                shape=(len(self.node_list), len(self.node_list)),
+            )
```
It looks like the new `StellarGraph` class (under development) should have a method to create adjacency matrices for different relation types.  This should be captured somewhere 📣

adocherty commented on stellargraph/mapper/full_batch_generators.py:
```diff
@@ -238,3 +238,209 @@ def flow(self, node_ids, targets=None):
+            if transform is None:
+                # normalize here and add 1e-9 row sum to avoid harmless divide by zero warnings
+                d = sps.diags(
+                    np.float_power(np.array(A.sum(1)) + 1e-9, -1).flatten(), 0
```
I don't particularly like adding a small number to solve DBZ problems. This depends on the float type of A, and the weights: it is fragile. Try this:

```
d = sps.diags(
    np.float_power(np.ravel(np.maximum(A.sum(axis=1),1)), -1), 0
)
```

kieranricardo replied:
much better solution than my `+1e-9`, fixed!

adocherty commented on stellargraph/mapper/full_batch_generators.py:
```diff
@@ -238,3 +238,209 @@ def flow(self, node_ids, targets=None):
+                d = sps.diags(
+                    np.float_power(np.array(A.sum(1)) + 1e-9, -1).flatten(), 0
+                )
+                A = d.dot(A).tocsr()
```
Why are you converting to CSR then on line 408 to COO?  After the dot it will most likely be CSR anyway.

kieranricardo replied:
nice catch, was a copy paste error. fixed!

adocherty commented on demos/node-classification/rgcn/rgcn-aifb-node-classification-example.ipynb:
```diff
@@ -0,0 +1,414 @@
+    "from networkx.classes.function import info\n",
+    "\n",
+    "import stellargraph as sg\n",
+    "from stellargraph.mapper.node_mappers import RelationalFullBatchNodeGenerator\n",
```
Don't import from `node_mappers` : this file no longer exists. The imports should be:
```
from stellargraph.mapper import RelationalFullBatchNodeGenerator
```

To do this you'll need to add `RelationalFullBatchNodeGenerator` to the `__all__` variable at the top of the `mappers/full_batch_generators.py` file so they are inserted into the `mappers` namespace.

kieranricardo replied:
done!

kieranricardo commented:
@PantelisElinas @adocherty thanks for your reviews! I think I've addressed all your comments + I've made things compatible with the new stellargaph class. All the units tests are passing but the build is failing (https://github.com/stellargraph/stellargraph/pull/562) so I think its ready for re-review

PantelisElinas commented on stellargraph/layer/rgcn.py:
```diff
@@ -0,0 +1,598 @@
+
+    For more details, please see the RGCN demo notebook:
+
+    emos/node-classification/rgcn/rgcn-aifb-node-classification-example.ipynb
```
I think you mean "demos/...", missing the letter d.

kieranricardo replied:
nice catch, I'll quickly fix that

adocherty commented on stellargraph/layer/rgcn.py:
```diff
@@ -480,7 +533,7 @@ def __call__(self, x):
         return h_layer
 
-    def node_model(self):
+    def _node_model(self):
```
Currently all our methods have `node_model` and `link_model` so it would be good to talk in the new team about if these methods in the other classes should be deprecated .. for now just be aware that this makes RGCN only have the new `build` API.

adocherty commented on stellargraph/mapper/full_batch_generators.py:
```diff
@@ -379,10 +309,14 @@ def __init__(self, G, name=None, sparse=True, transform=None):
+                self.node_index[n2]
+                for n1, n2, etype in G.edges(triple=True)
+                if etype == edge_type
             ]
```
Perhaps this functionality could be something for the new StellarGraph class?  Also have a look at HinSAGE and the `GraphSchema` class which was supposed to be a place for this functionality.  *There is no need to change this PR* but perhaps in the  future these operations can be done in the StellarGraph class so they can be optimized.
